flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/ExecutionGraphSignalsTest.java

```diff
@@ -0,0 +1,221 @@
+				any(FiniteDuration.class), any(Long.class).longValue()).thenReturn(mockEJV[0]);
+		PowerMockito
+			.whenNew(ExecutionJobVertex.class)
+			.withArguments(any(ExecutionGraph.class), same(v2), any(Integer.class).intValue(),
+				any(FiniteDuration.class), any(Long.class).longValue()).thenReturn(mockEJV[1]);
+		PowerMockito
+			.whenNew(ExecutionJobVertex.class)
+			.withArguments(any(ExecutionGraph.class), same(v3), any(Integer.class).intValue(),
+				any(FiniteDuration.class), any(Long.class).longValue()).thenReturn(mockEJV[2]);
+		PowerMockito
+			.whenNew(ExecutionJobVertex.class)
+			.withArguments(any(ExecutionGraph.class), same(v4), any(Integer.class).intValue(),
+				any(FiniteDuration.class), any(Long.class).longValue()).thenReturn(mockEJV[3]);
+		PowerMockito
+			.whenNew(ExecutionJobVertex.class)
+			.withArguments(any(ExecutionGraph.class), same(v5), any(Integer.class).intValue(),
+				any(FiniteDuration.class), any(Long.class).longValue()).thenReturn(mockEJV[4]);
+
+		v1.setParallelism(dop[0]);
+		v2.setParallelism(dop[1]);
+		v3.setParallelism(dop[2]);
+		v4.setParallelism(dop[3]);
+		v5.setParallelism(dop[4]);
+
+		v2.connectNewDataSetAsInput(v1, DistributionPattern.ALL_TO_ALL);
+		mockNumberOfInputs(1,0);
+		v4.connectNewDataSetAsInput(v2, DistributionPattern.ALL_TO_ALL);
+		mockNumberOfInputs(3,1);
+		v4.connectNewDataSetAsInput(v3, DistributionPattern.ALL_TO_ALL);
+		mockNumberOfInputs(3,2);
+		v5.connectNewDataSetAsInput(v4, DistributionPattern.ALL_TO_ALL);
+		mockNumberOfInputs(4,3);
+		v5.connectNewDataSetAsInput(v3, DistributionPattern.ALL_TO_ALL);
+		mockNumberOfInputs(4,2);
+
+		List<JobVertex> ordered = new ArrayList<JobVertex>(Arrays.asList(v1, v2, v3, v4, v5));
+
+		eg = new ExecutionGraph(TestingUtils.defaultExecutionContext(), jobId, jobName,
+				JobType.BATCHING, cfg, AkkaUtils.getDefaultTimeout());
+		eg.attachJobGraph(ordered);
+
+		f = eg.getClass().getDeclaredField("state");
+		f.setAccessible(true);
+	}
+
+	private void mockNumberOfInputs(int nodeIndex, int predecessorIndex) {
+		for(int j = 0; j < dop[nodeIndex]; ++j) {
+			when(mockEV[nodeIndex][j].getNumberOfInputs()).thenReturn(dop[predecessorIndex]);
+		}
+	}
+
+	@Test
+	public void testCancel() throws Exception {
+		Assert.assertEquals(JobStatus.CREATED, eg.getState());
+		eg.cancel();
+
+		verifyCancel(1);
+
+		f.set(eg, JobStatus.RUNNING);
+		eg.cancel();
+
+		verifyCancel(2);
+		Assert.assertEquals(JobStatus.CANCELLING, eg.getState());
+
+		eg.cancel();
+
+		verifyCancel(2);
+		Assert.assertEquals(JobStatus.CANCELLING, eg.getState());
+
+		f.set(eg, JobStatus.CANCELED);
+		eg.cancel();
+
+		verifyCancel(2);
+		Assert.assertEquals(JobStatus.CANCELED, eg.getState());
+
+		f.set(eg, JobStatus.FAILED);
+		eg.cancel();
+
+		verifyCancel(2);
+		Assert.assertEquals(JobStatus.FAILED, eg.getState());
+
+		f.set(eg, JobStatus.FAILING);
+		eg.cancel();
+
+		verifyCancel(2);
+		Assert.assertEquals(JobStatus.FAILING, eg.getState());
```

Comment:
I think that the transition from FAILING to CANCELLING is contra intuitive. But you made it very clear that this is the intended behavior. I will just update this test to check for `JobStatus.CANCELLING`.